### PR TITLE
Fixes #4658 - Use element input type text

### LINKED ISSUE #4658: [Annotations] [Texttool] Use text input instead of a contenteditable div
Refactor to switch from contenteditable div to input text to help fix issues open around page up/down/ text caret placement/undo/long strings 
#4646 
#4642 
#4632
#4655 
#4654 
#4653
#4652

### PR DESCRIPTION
**Pull request by punamdahiya**

### CODE REVIEW

**Comment by punamdahiya** [comment]
@chenba updating to use input type text does help fix below issues
#4646
#4642
#4632
#4655
#4654
#4653

However it loses the advantage of auto resize  text-input width as per the content and on confirm fails to identify text x position. Two approaches
a) Either fix adjustX logic inside CSS
b) Compute in JS line width by identifying width of a single character for respective font size??

File: server/src/pages/shot/text-tool.js [hunk 1]
```diff
@@ -73,10 +76,16 @@ exports.TextTool = class TextTool extends React.Component {
...
-  adjustHeight() {
+  adjustHeightAndWidth() {
     const styles = window.getComputedStyle(this.textInput.current);
     const height = Math.ceil(parseFloat(styles.lineHeight) + parseFloat(styles.paddingTop) + parseFloat(styles.paddingBottom));
     this.textInput.current.style.minHeight = `${height}px`;
+
+    inputTextSize = parseFloat(styles.fontSize);
+    inputTextPadding = parseFloat(styles.paddingLeft);
+
+    const width = Math.ceil(((inputTextSize / 2) * this.textInput.current.value.length) + parseFloat(styles.paddingLeft) + parseFloat(styles.paddingRight));
```
Comment (punamdahiya) [inline server/src/pages/shot/text-tool.js]: Assumption here single character width is fontSize / 2 is flawed resulting in incorrect x position on hit of confirm

File: server/src/pages/shot/text-tool.js [hunk 2]
```diff
@@ -228,41 +237,33 @@ exports.TextTool = class TextTool extends React.Component {
...
   }
 
   adjustX(e) {
     // Return if there's no text content change and method is invoked from event handler
-    if (previousInputText === this.textInput.current.textContent && e) {
+    if (previousInputText === this.textInput.current.value && e) {
       return;
     }
+
+    const newWidth = Math.ceil(((inputTextSize / 2) * (this.textInput.current.value.length + 1)) + (2 * inputTextPadding));
```
Comment (punamdahiya) [inline server/src/pages/shot/text-tool.js]: same as above comment

**Comment by chenba** [comment]
Have you tried adjusting the width of the input by updating its `size` attribute?

**Comment by punamdahiya** [comment]
Trying size is changing visible width of input since it's measured in characters , for 36px font-size it's showing width as 72.5 leaving extra space on both sides for text-align center.

**Comment by chenba** [comment]
Could you hide a div that has the same text content as the input so you can get the width from the div?

**Comment by punamdahiya** [comment]
Yes, getting width of the text content using a hidden div should work

**Comment by chenba** [comment]
Fixes #4658.  (Adding this so that the issue is linked to this PR.)

**Comment by punamdahiya** [comment]
@chenba Updated PR to use hidden div to compute input text width and is ready for review. Thanks!

File: server/src/pages/shot/text-tool.js [hunk 3]
```diff
@@ -119,6 +132,11 @@ exports.TextTool = class TextTool extends React.Component {
     prevDragMousePos = this.captureMousePosition(e);
   }
 
+  onDrop(e) {
+    // Block drop event inside text input
+    e.preventDefault();
+  }
```
Comment (chenba) [inline server/src/pages/shot/text-tool.js]: I don't think we need this now that it's a text `input`.

Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: We need this for stopping drop of formatted text from a text editor as shown in #4662. Apparently text input allows it

File: server/src/pages/shot/text-tool.js [hunk 4]
```diff
@@ -102,11 +105,21 @@ exports.TextTool = class TextTool extends React.Component {
       cursor: "move"
     };
 
+    // Styles used by hidden div that is used to compute width of input
+    // text for respective font-size
+    const hiddenDivStyles = {
+      position: "absolute",
+      "z-index": "-999",
```
Comment (chenba) [inline server/src/pages/shot/text-tool.js]: JSX accepts `zIndex` instead of `z-index`.

File: server/src/pages/shot/text-tool.js (hunk 3 above)
Comment (chenba) [reply server/src/pages/shot/text-tool.js]: But the dropped content is not formatted and doesn't break anything.  It's nice being able to drag some text into the input.

Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: it sure work now when I tried :)

**Comment by punamdahiya** [comment]
@chenba updated to use onInput, the changes are in separate commit for easy review.

File: server/src/pages/shot/text-tool.js [hunk 5]
```diff
@@ -102,11 +105,21 @@ exports.TextTool = class TextTool extends React.Component {
       cursor: "move"
     };
 
+    // Styles used by hidden div that is used to compute width of input
+    // text for respective font-size
+    const hiddenDivStyles = {
+      position: "absolute",
+      zIndex: "-999",
+      opacity: "0",
+      top: "-999px"
```
Comment (chenba) [inline server/src/pages/shot/text-tool.js]: We should add `visibility: hidden` here.

File: server/src/pages/shot/text-tool.js [hunk 6]
```diff
@@ -73,10 +73,13 @@ exports.TextTool = class TextTool extends React.Component {
     }
   }
 
-  adjustHeight() {
+  adjustHeightAndWidth() {
     const styles = window.getComputedStyle(this.textInput.current);
     const height = Math.ceil(parseFloat(styles.lineHeight) + parseFloat(styles.paddingTop) + parseFloat(styles.paddingBottom));
     this.textInput.current.style.minHeight = `${height}px`;
```
[the remark this answers is not in the document]
Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: +1

File: server/src/pages/shot/text-tool.js (hunk 5 above)
Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: with visibility:hidden, we can safely remove opacity and top here

File: server/src/pages/shot/text-tool.js [hunk 7]
```diff
@@ -228,41 +241,28 @@ exports.TextTool = class TextTool extends React.Component {
...
-      this.adjustX(e);
-    });
-  }
-
-  onKeyUp(e) {
-    // Fix to remove <br> element inserted on press of space bar inside contenteditable div
-    while (this.textInput.current.firstElementChild) {
-      this.textInput.current.removeChild(this.textInput.current.firstElementChild);
-    }
+  onInput(e) {
```
Comment (chenba) [inline server/src/pages/shot/text-tool.js]: Can we get rid of `onKeyDown` then?

Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: need onkeydown to handle arrow keys when no input is there

Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: or is it handled by text input handlers? will check

Comment (punamdahiya) [reply server/src/pages/shot/text-tool.js]: yes, looks good. Should be ok to remove

**Comment by punamdahiya** [comment]
Updated PR to remove adjust height and keydown handler

**Pull request merged**